### PR TITLE
Fixed worker crashing with bad __repr__ for args

### PR DESCRIPTION
**Pull request by amosonn**
Previously, when a worker job failed, it tried to log the args and kwargs which led to a failure. In order to do so, it invoked the __repr__ of said arguments (indirectly, via __str__ of args and kwargs), and if those were broken, crashed.
Now it try..excepts these castings, and puts a stock value when they fail.
Includes test.

### CODE REVIEW

File: distributed/worker.py [hunk 1]
```diff
@@ -760,3 +763,41 @@ def apply_function(function, args, kwargs):
...
+    """ Convert args to a string, allowing for some arguments to raise
+    exceptions during conversion and ignoring them.
+    """
+    sargs = "("
+    for arg in args:
+        try:
+            sarg = repr(arg)
+        except:
+            sarg = "< could not convert arg to str >"
+        sargs += sarg + ", "
```
Comment (mrocklin) [inline distributed/worker.py]: It is slightly more efficient if you make a list of strings and then create the result in one go with `join` as in `'(' + ', '.join(strings) + ')'` instead.  Otherwise we create many copies.  This isn't a big deal though (number of args is usually small).

File: distributed/tests/test_worker.py [hunk 2]
```diff
@@ -55,6 +55,84 @@ def test_health():
...
+                                    who_has={'x': [a.address]})
+        assert not b.active
+        assert response['status'] == 'error'
+        assert isinstance(loads(response['exception']), ZeroDivisionError)
+        if sys.version_info[0] >= 3:
+            assert any('1 / 0' in line
+                      for line in pluck(3, traceback.extract_tb(
+                          loads(response['traceback'])))
+                      if line)
+        # TODO: test log item.
```
Comment (mrocklin) [inline distributed/tests/test_worker.py]: You could either redirect the `logger` using the logging module or else use the `capsys` pytest fixture to capture stderr.  See `distributed/diagnostics/tests/test_progress.py:def check_bar_completed` for an example.

File: distributed/worker.py (hunk 1 above)
Comment (amosonn) [inline distributed/worker.py]: My consideration was that we want to lazily repr() things, so if we run out of character limit we don't repr redundant things, I guess this can be achieved with length arithmetic as well. I will rewrite.

File: distributed/tests/test_worker.py (hunk 2 above)
Comment (amosonn) [inline distributed/tests/test_worker.py]: will check.

File: distributed/worker.py (hunk 1 above)
Comment (mrocklin) [inline distributed/worker.py]: Oh sure, that makes sense.  Do as you think is best.

**Comment by amosonn** [comment]
I added some code snippet to test the logging. It seems the right tool for the job is testfixtures, see https://pythonhosted.org/testfixtures/logging.html
I of course leave at your discretion to choose whether to adopt it; since there is quite a lot of logging in distributed, I think good logging tests could go a long way.

File: distributed/tests/test_worker.py [hunk 3]
```diff
@@ -55,6 +56,109 @@ def test_health():
         assert 'network-send' in d
 
 
+def test_worker_bad_args(loop, capsys):
```
Comment (mrocklin) [inline distributed/tests/test_worker.py]: Can probably drop `capsys` with your `MockLoggingHandler`

File: distributed/tests/test_worker.py [hunk 4]
```diff
@@ -55,6 +56,109 @@ def test_health():
...
+        if sys.version_info[0] >= 3:
+            assert any('1 / 0' in line
+                      for line in pluck(3, traceback.extract_tb(
+                          loads(response['traceback'])))
+                      if line)
+        assert hdlr.messages['warning'][0] == " Compute Failed\n" \
+            "Function: bad_func\n" \
+            "args:     (< could not convert arg to str >)\n" \
+            "kwargs:   {'k': < could not convert arg to str >}\n"
+
```
Comment (mrocklin) [inline distributed/tests/test_worker.py]: Are the lines below this comment relevant for this change?

File: distributed/tests/test_worker.py [hunk 5]
```diff
@@ -55,6 +56,109 @@ def test_health():
         assert 'network-send' in d
 
 
+def test_worker_bad_args(loop, capsys):
+    @gen.coroutine
+    def f(c, a, b):
```
Comment (mrocklin) [inline distributed/tests/test_worker.py]: You may be interested in switching this out for a slightly newer testing style

```python
@gen_cluster
def test_worker_bad_str_args(c, a, b):
    ...
```

See test_executor.py for examples.  I should probably rewrite the tests in this file in the new style to avoid issues like this.

Feel free to ignore.  I can handle this later.

**Comment by mrocklin** [comment]
testfixtures looks cool.  A quick google search also yields https://pypi.python.org/pypi/pytest-capturelog 

This deserves some thought.

File: distributed/tests/test_worker.py (hunk 4 above)
Comment (amosonn) [inline distributed/tests/test_worker.py]: the exception checks are copy-pasted from the other test, but I wanted to leave them to make sure it is indeed the division-by-zero that break the job, not the failure to repr().
The last assert is the check that indeed the logging worked.

Comment (mrocklin) [inline distributed/tests/test_worker.py]: I think that you're referring to the lines above this comment.  The lines below this comment seem to be unrelated?  I could be confused here.

**Comment by amosonn** [comment]
Yeah, I don't have much to say about either, no experience with testing logging either way :). But it seems the capturelog is easier to use and also more pytest ready.

File: distributed/tests/test_worker.py (hunk 4 above)
Comment (amosonn) [inline distributed/tests/test_worker.py]: Ah, sorry, I thought it was odd that you wrote above, but these were the lines that were shown next to your comment.
The lines below are to check that none of the workers died as a result of processing the sad object - so they're still accepting jobs and processing them normally. Perhaps a comment is in place. I will add.

**Comment by mrocklin** [comment]
This looks good to me.  The errors that we're seeing confuse me a bit.  They don't appear to be related.

**Comment by amosonn** [comment]
Yeah, they also come and go: the last build succeeded (before adding the
comments). One of them is sort of a timeout, so maybe increasing that
timeout will solve it, and I have a hunch that the other is related.

**Comment by mrocklin** [comment]
Verified that the `test_bad_tasks_fail` test is unrelated.  Feel free to bump the timeout to 20 seconds.  I'll do this in another branch soon as well.

**Comment by amosonn** [comment]
Increasing the timeouts indeed solves test_bad_tasks_fail.
However, test_broken_worker_during_computation also fails, sometimes, and I haven't managed to understand why. But it still seems unrelated.

**Comment by amosonn** [comment]
Seems that increasing the timeout still doesn't really help. So probably something is amiss. What do you recommend next?

**Comment by mrocklin** [comment]
I've run into this on my own work and am handling at least some of it.  I recommend that we just let this PR sit tight for a couple of days after I've resolved some internal issues.

**Comment by amosonn** [comment]
Cool, so I'll throw away the false timeout fixes.

If you have some pointer to throw me at I can try to give it a shot.

**Comment by mrocklin** [comment]
I've merged in a fairly large change.  I apologize that this caused some merge conflicts on your branch.  I've gone through the rebase on by bad-worker-args if you'd like to reset to the changes.  Alternatively I can just merge from there.

**Comment by mrocklin** [comment]
Ah, the branch name is `worker-bad-args`

```
git remote add mrocklin git@github.com:mrocklin/distributed.git
git remote update mrocklin
git reset --hard mrocklin/worker-bad-args
```

https://github.com/mrocklin/distributed/tree/worker-bad-args

**Comment by amosonn** [comment]
Thanks!
I wasn't sure exactly what happened, so I've just rebased it myself on top of master.

**Comment by amosonn** [comment]
However, the logging tests now fail on travis, even though they still pass on my machine. Should I just drop them for now?

**Comment by mrocklin** [comment]
Thanks @amosonn !  This is in.
